src/sinks/util/acker_bytes_sink.rs

```diff
@@ -0,0 +1,73 @@
+use crate::buffers::Acker;
+use bytes::Bytes;
+use futures::{ready, Sink};
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::AsyncWrite;
+use tokio_util::codec::{BytesCodec, FramedWrite};
+
+const MAX_PENDING_ITEMS: usize = 10_000;
+
+#[pin_project]
+pub struct AckerBytesSink<T> {
+    #[pin]
+    inner: FramedWrite<T, BytesCodec>,
+    sizes: Vec<usize>,
+    acker: Acker,
+    on_success: Box<dyn Fn(usize) + Send>,
+}
+
+impl<T: AsyncWrite> AckerBytesSink<T> {
+    pub fn new(inner: T, acker: Acker, on_success: Box<dyn Fn(usize) + Send>) -> Self {
+        Self {
+            inner: FramedWrite::new(inner, BytesCodec::new()),
+            sizes: Vec::with_capacity(MAX_PENDING_ITEMS),
+            acker,
+            on_success,
+        }
+    }
+
+    pub fn ack(&mut self) {
+        self.acker.ack(self.sizes.len());
+        for size in self.sizes.drain(..) {
+            (self.on_success)(size);
+        }
+    }
+}
+
+impl<T> Sink<Bytes> for AckerBytesSink<T>
+where
+    T: AsyncWrite + std::marker::Unpin,
+{
+    type Error = <FramedWrite<T, BytesCodec> as Sink<Bytes>>::Error;
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if self.as_mut().project().sizes.len() == MAX_PENDING_ITEMS {
+            if let Err(error) = ready!(self.as_mut().poll_flush(cx)) {
+                return Poll::Ready(Err(error));
+            }
+        }
+        self.project().inner.poll_ready(cx)
+    }
+
+    fn start_send(self: Pin<&mut Self>, item: Bytes) -> Result<(), Self::Error> {
+        let pinned = self.project();
+        pinned.sizes.push(item.len());
+        pinned.inner.start_send(item)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
```

Comment:
This is probably just my ignorance, but can you explain how the failure case works here? What happens if, for example, the unix socket cannot send the event in `call` below?

Reply:
If `UnixSocket` cannot send the event then `FramedWrite` will pass error to us (to `result` variable), but we do not have any logic here for handling this error so we also pass it on next level. On next level [SendAll::poll](https://docs.rs/futures-util/0.3.6/src/futures_util/sink/send_all.rs.html#78-112) also pass error to our loop in `StreamSink::run` method and finally there we handler error (by emitting event and reconnecting).